text/0000-app-mixins.md

```diff
@@ -0,0 +1,130 @@
+- Author(s): [Joe Kutner](https://github.com/jkutner/)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for allowing application developers (buildpack users) to specify mixins that will be dynamically included with their build and run images. In this way, users can include custom system packages (and other libraries) with their applications.
+
+# Motivation
+[motivation]: #motivation
+
+Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time, but it's common for application code to depend on OS packages too. But allowing buildpacks to arbitrarily install OS packages during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements.
+
+For that reason, this proposal defines a mechanism to implement dynamic installation of mixins at build time while minimizing the performance impact. This is accomplished by allowing stack images to satisfy mixin requires statically and/or reject dynamic mixin installation early in the build process.
+
+# What it is
+[what-it-is]: #what-it-is
+
+- *mixin* - a named set of additions to a stack that can be used to make additive changes to the contract.
+- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
+- *stack buildpack* - a type of buildpack that runs against the stack image(s) instead of an app. Among other things, it can provide mixins.
+
+An application developer may specify a list of mixins in their application's `project.toml` file like this:
+
+```toml
+[build]
+mixins = [ "<mixin name>" ]
+```
+
+When a command like `pack build` is run, the list of mixins will be processed before buildpacks are run. For each mixin name, the following will happen:
+
+* If the mixin name is prefixed with `build:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the build image.
+* If the mixin name is prefixed with `run:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the run image.
+* If the mixin name does not have a prefix it will be dynamically added to both the build and run stack images.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When a list of mixins are required by buildpacks via the build plan and the build phase starts:
+
+1. The lifecycle will compare the list of mixins to those provided by the stack using the `io.buildpacks.stack.mixins` label.
+1. If all mixin names are provided by the stack, no further action is required.
+1. If any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.
```

Comment:
I'm a bit confused - does this mean that we run `detect` twice? I thought stack packs were supposed to run before userspace buildpacks...